ubi/src/installer.rs

```diff
@@ -177,6 +181,64 @@ impl ExeInstaller {
         Ok(possible_matches.into_iter().next())
     }
 
+    fn extract_executable_from_7z(&self, downloaded_file: &Path) -> Result<PathBuf> {
+        debug!(
+            "extracting executable from 7z file at {}",
+            downloaded_file.display()
+        );
+
+        let mut archive = ArchiveReader::new(open_file(downloaded_file)?, Password::empty())?;
+
+        if let Some(file_name) = self.best_match_from_7z_archive(&archive) {
+            let mut install_path = self.install_path.clone();
+            if let Some(ext) = Extension::from_path(&PathBuf::from(&file_name))? {
+                if ext.should_preserve_extension_on_install() {
+                    debug!("preserving the {} extension on install", ext.extension());
+                    install_path.set_extension(ext.extension_without_dot());
+                }
+            }
+
+            debug!(
+                "extracting 7z entry named {} to {}",
+                file_name,
+                install_path.display(),
+            );
+            let buffer = archive.read_file(&file_name)?;
+            self.create_install_dir()?;
+
+            File::create(&install_path)?.write_all(&buffer)?;
+
+            return Ok(install_path);
+        }
+
+        self.could_not_find_archive_matches_error()
+    }
+
+    fn best_match_from_7z_archive(&self, archive: &ArchiveReader<File>) -> Option<String> {
+        let mut possible_matches: Vec<String> = vec![];
+
+        for file in &archive.archive().files {
+            if file.is_directory() {
+                continue;
+            }
+
+            let path = PathBuf::from(file.name());
+            if let Some(file_name) = path.file_name() {
+                if let Some(file_name) = file_name.to_str() {
+                    if self.archive_member_is_exact_match(file_name) {
+                        debug!("found 7z file entry with exact match: {file_name}");
+                        return Some(file.name().to_string());
+                    } else if self.archive_member_is_partial_match(file_name) {
+                        debug!("found 7z file entry with partial match: {file_name}");
+                        possible_matches.push(file.name().to_string());
+                    }
+                }
+            }
+        }
+
+        possible_matches.into_iter().next()
+    }
```

Comment:
I wonder if it'd be possible to combine this logic with the very similar logic for tarballs and zip files. It seems like the only thing that differs is the way you iterate over files in an archive and the way you get the filename from the entry.

I think the best way to handle this would be to add two new traits, one for iteration and one for getting a PathBuf from an iterator. Then we could implement that trait for the tar, zip, & 7z readers, as well as the entries that iteration returns.

I will work on this for the existing tarball and zip readers, then you can implement it for the 7z archives too.

Reply:
I just merged that branch.